File: ports/go/stores/mcp_test.go
package stores

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/jagentic/goagentic/core"
)

const stubMCPServer = `
from mcp.server.fastmcp import FastMCP
mcp = FastMCP("stub")

@mcp.tool()
def add(a: int, b: int) -> int:
    """Add two numbers."""
    return a + b

if __name__ == "__main__":
    mcp.run()
`

func TestMcpClientRegistersAndCallsToolsIfAvailable(t *testing.T) {
	py := os.Getenv("AGENTIC_TEST_PYTHON")
	if py == "" {
		py = "/tmp/af-venv/bin/python"
	}
	if _, err := os.Stat(py); err != nil {
		t.Skipf("python with mcp SDK not available at %s", py)
	}
	stub := filepath.Join(t.TempDir(), "stub_mcp_server.py")
	if err := os.WriteFile(stub, []byte(stubMCPServer), 0o644); err != nil {
		t.Fatal(err)
	}
	client, err := NewMcpClient(py, nil, stub)
	if err != nil {
		t.Skipf("MCP stub server didn't start (mcp SDK missing?): %v", err)
	}
	defer client.Close()

	names := map[string]bool{}
	for _, s := range client.Tools() {
		names[s["name"]] = true
	}
	if !names["add"] {
		t.Fatalf("expected 'add' tool, got %v", client.Tools())
	}
	reg := core.NewToolRegistry()
	ids := client.Register(reg, "mcp_")
	if len(ids) == 0 || !reg.Has("mcp_add") {
		t.Fatalf("register ids = %v", ids)
	}
	got := strings.TrimSpace(reg.Execute("mcp_add", map[string]any{"a": 2, "b": 3}).(string))
	if got != "5" {
		t.Fatalf("mcp_add(2,3) = %q, want 5", got)
	}
}

File: ports/jagentic-core/src/test/java/org/jagentic/core/McpTest.java
package org.jagentic.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.Test;

import org.jagentic.core.store.McpStdioClient;

/** Phase D — pure-Java MCP stdio client against a tiny FastMCP stub server (venv python).
 * Skips cleanly when the python+mcp SDK isn't present. */
class McpTest {

  private static final String STUB_SERVER = """
      from mcp.server.fastmcp import FastMCP
      mcp = FastMCP("stub")

      @mcp.tool()
      def add(a: int, b: int) -> int:
          \"\"\"Add two numbers.\"\"\"
          return a + b

      @mcp.tool()
      def shout(text: str) -> str:
          \"\"\"Uppercase the text.\"\"\"
          return text.upper()

      if __name__ == "__main__":
          mcp.run()
      """;

  @Test
  void mcpClientRegistersAndCallsTools() throws IOException {
    String py = System.getenv("AGENTIC_TEST_PYTHON");
    if (py == null || py.isBlank()) {
      py = "/tmp/af-venv/bin/python";
    }
    if (!Files.exists(Path.of(py))) {
      Assumptions.abort("python with mcp SDK not available at " + py);
      return;
    }
    Path stub = Files.createTempFile("stub_mcp_server", ".py");
    Files.writeString(stub, STUB_SERVER);

    McpStdioClient client;
    try {
      client = new McpStdioClient(List.of(py, stub.toString()));
    } catch (Throwable t) {
      Assumptions.abort("MCP stub server didn't start (mcp SDK missing?): " + t.getMessage());
      return;
    }
    try {
      Set<String> names = client.tools().stream()
          .map(s -> s.get("name")).collect(Collectors.toSet());
      assertTrue(names.containsAll(Set.of("add", "shout")), "tools: " + names);

      ToolRegistry reg = new ToolRegistry();
      List<String> ids = client.register(reg, "mcp_");
      assertTrue(ids.contains("mcp_add"));
      assertEquals("5", ((String) reg.execute("mcp_add", Map.of("a", 2, "b", 3))).trim());
      assertEquals("HI", ((String) reg.execute("mcp_shout", Map.of("text", "hi"))).trim());
    } finally {
      client.close();
    }
  }
}
